src/int.rs

```diff
@@ -0,0 +1,1409 @@
+    /// whether an arithmetic overflow would occur. If an overflow would have
+    /// occurred then the wrapped value is returned.
+    pub fn overflowing_mul(self, rhs: Self) -> (Self, bool) {
+        let sign = Sign::from_signum64(self.signum64() * rhs.signum64());
+        let (unsigned, overflow_mul) = self.abs_unsigned().overflowing_mul(rhs.abs_unsigned());
+        let (result, overflow_conv) = I256::overflowing_from_sign_and_abs(sign, unsigned);
+
+        (result, overflow_mul || overflow_conv)
+    }
+
+    /// Checked multiplication. Returns None if overflow occurred.
+    pub fn checked_mul(self, other: Self) -> Option<Self> {
+        let (result, overflow) = self.overflowing_mul(other);
+        if overflow {
+            None
+        } else {
+            Some(result)
+        }
+    }
+
+    /// Multiplication which saturates at the maximum value..
+    pub fn saturating_mul(self, rhs: Self) -> Self {
+        self.checked_mul(rhs).unwrap_or_else(|| {
+            match Sign::from_signum64(self.signum64() * rhs.signum64()) {
+                Sign::Positive => I256::MAX,
+                Sign::Negative => I256::MIN,
+            }
+        })
+    }
+
+    /// Wrapping multiplication.
+    pub fn wrapping_mul(self, rhs: Self) -> Self {
+        let (result, _) = self.overflowing_mul(rhs);
+        result
+    }
+
+    /// Returns the sign of `self` to the exponent `exp`.
+    ///
+    /// Note that this method does not actually try to compute the `self` to the
+    /// exponent `exp`, but instead uses the property that a negative number to
+    /// an odd exponent will be negative. This means that the sign of the result
+    /// of exponentiation can be computed even if the actual result is too large
+    /// to fit in 256-bit signed integer.
+    fn pow_sign(self, exp: u32) -> Sign {
+        let is_exp_odd = exp % 2 != 0;
+        if is_exp_odd && self.is_negative() {
+            Sign::Negative
+        } else {
+            Sign::Positive
+        }
+    }
+
+    /// Create 10**n as this type.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result overflows the type.
+    pub fn exp10(n: usize) -> Self {
+        U256::exp10(n).try_into().expect("overflow")
+    }
+
+    /// Raise self to the power of `exp`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result overflows the type in debug mode.
+    pub fn pow(self, exp: u32) -> Self {
+        handle_overflow(self.overflowing_pow(exp))
+    }
+
+    /// Raises self to the power of `exp`.
+    ///
+    /// Returns a tuple of the exponentiation along with a bool indicating
+    /// whether an overflow happened.
+    pub fn overflowing_pow(self, exp: u32) -> (Self, bool) {
+        let sign = self.pow_sign(exp);
+        let (unsigned, overflow_pow) = self.abs_unsigned().overflowing_pow(exp.into());
+        let (result, overflow_conv) = I256::overflowing_from_sign_and_abs(sign, unsigned);
+
+        (result, overflow_pow || overflow_conv)
+    }
```

Comment:
I need to understand the overflowing better... At first I assumed the overflowing power would be the one that wraps around, but then I see you have a `wrapping_pow` below. This can only mean that this overflowing power actually gives the correct result in the event that the numbers we unbounded. However, this would mean that the result could fall out of scope and I would no longer know what type to expect... 😕 ❓

Reply:
- `overflowing_*` methods compute the result of the operation and return the value and returns both the value, that is wrapped in case of an overflow, and a bool indicating whether or not an overflow occured.
- `wrapping_*` return the exact same result value, but just ignore the overflow bool.
- `checked_*` retruns `Some(value)` when the result fits in the signed integer (i.e. no overflow) and `None` if there is an overflow.
- `saturating_*` returns the value result of the operation if the result fits in the signed integer and `I256::MAX` or `I256::MIN` depending on the direction of the overflow. 

Hope this explains it a little better.